cpp/src/arrow/builder.h

```diff
@@ -199,12 +199,16 @@ class ARROW_EXPORT PrimitiveBuilder : public ArrayBuilder {
   /// Write nulls as uint8_t* (0 value indicates null) into pre-allocated memory
   Status AppendNulls(const uint8_t* valid_bytes, int64_t length) {
     RETURN_NOT_OK(Reserve(length));
+    memset(raw_data_ + length_, 0,
```

Comment:
Wait... If `valid_bytes` contains non-null entries, and I already filled in those non-null entries, this will overwrite them with zeros?
I think at the very least this should be mentioned in the docstring.

Reply:
Yeah, I wasn't sure about the intended semantics of `AppendNulls`. Looking at the only use case in the orc adapter, it doesn't cause problems, because it's called before the data is set, but I see how it can be a nasty surprise.

The problem is that we have no way of knowing if the data at the non-null place is uninitialized or intentionally set, so I'll go with a warning in a comment.